packages/zoe/src/contracts/vpool-xyk-amm/multipoolMarketMaker.js

```diff
@@ -0,0 +1,193 @@
+import { makeDoublePool } from './doublePool';
+
+/**
+ * Multipool AMM is a rewrite of Uniswap that supports multiple liquidity pools,
+ * and direct exchanges across pools. Please see the documentation for more:
+ * https://agoric.com/documentation/zoe/guide/contracts/multipoolAMM.html It
+ * also uses a unique approach to charging fees. Each pool grows on every trade,
+ * and a protocolFee is also extracted.
+ *
+ * We expect that this contract will have tens to hundreds of issuers.  Each
+ * liquidity pool is between the central token and a secondary token. Secondary
+ * tokens can be exchanged with each other, but only through the central
+ * token. For example, if X and Y are two token types and C is the central
+ * token, a swap giving X and wanting Y would first use the pool (X, C) then the
+ * pool (Y, C). There are no liquidity pools directly between two secondary
+ * tokens.
+ *
+ * There should only need to be one instance of this contract, so liquidity can
+ * be shared as much as possible.
+ *
+ * When the contract is instantiated, the central token is specified in the
+ * terms. Separate invitations are available by calling methods on the
+ * publicFacet for adding and removing liquidity and for making trades. Other
+ * publicFacet operations support querying prices and the sizes of pools. New
+ * Pools can be created with addPool().
+ *
+ * When making trades or requesting prices, the caller must specify either a
+ * maximum input amount (swapIn, getInputPrice) or a minimum output amount
+ * (swapOut, getOutPutPrice) or both. For swaps, the required keywords are `In`
+ * for the trader's `give` amount, and `Out` for the trader's `want` amount.
+ * getInputPrice and getOutputPrice each take two Amounts. The price functions
+ * return both amountIn (which may be lower than the original amount) and
+ * amountOut (which may be higher). When both prices are specified, no swap will
+ * be made (and no price provided) if both restrictions can't be honored.
+ *
+ * When adding and removing liquidity, the keywords are Central, Secondary, and
+ * Liquidity. adding liquidity has Central and Secondary in the `give` section,
+ * while removing liquidity has `want` and `give` swapped.
+ *
+ * Transactions that don't require an invitation include addPool and the
+ * queries: getInputPrice, getOutputPrice, getPoolAllocation,
+ * getLiquidityIssuer, and getLiquiditySupply.
+ *
+ * @type {ContractStartFn}
+ */
+const start = zcf => {
+  // This contract must have a "Central" keyword and issuer in the
+  // IssuerKeywordRecord.
+  // poolFee is the portion of the fees that go into the pool (in BP).
+  // protocolFee is the portion of the fees that are shared with validators
+  const {
+    brands: { Central: centralBrand },
+    timer,
+    poolFee,
+    protocolFee,
+  } = zcf.getTerms();
+  assertIssuerKeywords(zcf, ['Central']);
+  assert(centralBrand !== undefined, X`centralBrand must be present`);
```

Comment:
A contract relies on its terms, yes? That is: it is in no position to defend against an attacker that provides terms maliciously?

The reason I ask is: this assertion doesn't actually establish that `centralBrand` is a brand (for example, by asking a trusted issuer). And it's passed to other functions whose argument type is `Brand`. My first thought was to change the type to `unknown` to see if any of the code actually relies on anything other than the identity of `centralBrand`, but then I realized that terms don't come from clients. They're endowed by a creator.

I don't see anything about this in the [getTerms API docs](https://agoric.com/documentation/zoe/api/zoe-contract-facet.html#zcf-getterms).